### PR TITLE
Apply chromatic adaptation between D50 and D65 standard referents

### PR DESCRIPTION
**Pull request by danburzo**
I've added an example implementation for the chromatic adaptation described in #42. I still need to fix the tests after I find a reliable source to check against. 

(I've also added `tap-spec` to the dev dependencies, which makes the test output look nicer)

### CODE REVIEW

**Comment by danburzo** [comment]
Also, since we only use the XYZ space as an intermediary between RGB and Lab, to my mind we could pre-multiply the XYZ ↔ RGB and D50 ↔ D65 matrices and simplify the code, with the price of a more "obscure" matrix.

**Comment by danburzo** [comment]
And I think the matrices may need adjustment, as described in this issue: https://github.com/w3c/csswg-drafts/issues/2492 :/

File: test/hcl-test.js [hunk 1]
```diff
@@ -13,17 +13,17 @@ tape("hcl(…) returns an instance of hcl and color", function(test) {
...
 });
 
 tape("hcl(…) returns defined hue and chroma, even for black and white", function(test) {
   test.hclEqual(color.hcl("black"), 0, 0, 0, 1);
   test.hclEqual(color.hcl("#000"), 0, 0, 0, 1);
   test.hclEqual(color.hcl(color.lab("#000")), 0, 0, 0, 1);
-  test.hclEqual(color.hcl("white"), 158.19859051364818, 0.00001795054880958058, 100.00000386666655, 1);
-  test.hclEqual(color.hcl("#fff"), 158.19859051364818, 0.00001795054880958058, 100.00000386666655, 1);
-  test.hclEqual(color.hcl(color.lab("#fff")), 158.19859051364818, 0.00001795054880958058, 100.00000386666655, 1);
+  test.hclEqual(color.hcl("white"), 0, 0.003457073518231546, 100, 1);
```
Comment (danburzo) [inline test/hcl-test.js]: Since white is `1 0 0` in Lab, `0` is the appropriate hue in HCL (`Math.atan2(b,a)`).

**Comment by danburzo** [comment]
I've fixed the HCL/Lab tests, where I've seen a slight change in the numbers, with the one surprise on HCL('white'), which now offers a more straightforward value.

The tests can be verified in Matlab, with:

* [`rgb2lab([r g b], 'WhitePoint', 'd50')`](https://www.mathworks.com/help/images/ref/rgb2lab.html) where `r`, `g`, and `b` are `[0..1]`.
* [`rgb2lab([L a b], 'WhitePoint', 'd50')`](https://www.mathworks.com/help/images/ref/lab2rgb.html)

by (ab)using the interactive examples. I've tested a few numbers and they seem to match.

**Comment by danburzo** [comment]
In regards to the matrices: after clarifying that the ICC standard uses an inaccurate D50 white point, I went for the standard D50 definition and used the direct _linear RGB_ ↔ _XYZ<sub>D50</sub>_ matrices from [Bruce Lindbloom's website](http://www.brucelindbloom.com/index.html?Eqn_RGB_to_XYZ.html).

**Comment by danburzo** [comment]
P.S. I'd like to check first the effect of the changes to the LAB / HCL color spaces to d3-interpolate and other color scale packages that depend on d3-color

**Comment by danburzo** [comment]
I'll try to get some reference values from one of them, or maybe http://www.colour-science.org/.

**Comment by danburzo** [comment]
I made an Observable notebook to compare the interpolation in three libraries: https://beta.observablehq.com/@danburzo/culori-scales-test

(Culori uses the same chromatic adaptation as this pull request).

There are some problems around interpolating in LCH with achromatic colors, related to the hue of white. In the previous implementation it was 158.2, and in the current one it's zero.

Chroma just puts NaN and (presumably) assumes zero-distance between hue.

**Comment by mbostock** [comment]
Here’s a notebook I made to compare the behavior of Chromatist and d3-color:

https://beta.observablehq.com/@mbostock/chromatist-vs-d3-color

**Comment by danburzo** [comment]
For the current implementation of d3-color (with D65 illuminant) I get ([script here](https://github.com/danburzo/color-test/blob/master/d65.js)):

```
maximum LAB -> sRGB error: 0.005395595768882623
maximum LAB -> sRGB (clamped) error: 1
maximum sRGB -> LAB error: 0.013983545325596024
```

And for the D50 implementation ([script here](https://github.com/danburzo/color-test/blob/master/d50.js)):

```
maximum LAB -> sRGB error: 0.0025086315553252106
maximum LAB -> sRGB (clamped) error: 1
maximum sRGB -> LAB error: 0.02080372021691621
```

The "clamped" error compares the resulting RGBs when converted to [0, 255] and in both implementations it matches python colour  with at most a variation of ±1 on a single R/G/B channel.

**Comment by danburzo** [comment]
(In regards to the tangential question on LCH interpolation, I've [moved it to d3-interpolate](https://github.com/d3/d3-interpolate/issues/50))

**Comment by mbostock** [comment]
It would be “cheating”, but we could also consider checking for achromatic colors in RGB space (r = g = b) before converting to Lab, and set a = b = 0:

https://github.com/d3/d3-color/blob/3806fc588a275d688ff1741b41e70cc2250efcf5/src/lab.js#L14-L28

Then when converting from Lab to Hcl, if a = b = 0, set h = NaN instead of 0:

https://github.com/d3/d3-color/blob/3806fc588a275d688ff1741b41e70cc2250efcf5/src/lab.js#L80-L85

**Comment by mbostock** [comment]
In re. to previous comment, proposed #45.

File: test/hcl-test.js (hunk 1 above)
Comment (svgeesus) [reply test/hcl-test.js]: White is 100 0 0 in Lab

**Comment by svgeesus** [comment]
> Chroma puts H: NaN if C = 0, allowing it to not interpolate on the H, whose value is I think of no importance for achromatic colors.

Sounds like a case for atan2 rather than atan (avoid the divide by zero trap)

**Comment by svgeesus** [comment]
> It would be “cheating”, but we could also consider checking for achromatic colors in RGB space (r = g = b) before converting to Lab, and set a = b = 0:

That does assume that the screen produces perfectly aligned ramps; which tends to only be true on a calibrated screen when adjustments to the video LUT have been made to ensure this.

**Comment by mbostock** [comment]
Merged with #45 into #46.

**Pull request closed without merging**